query-algebrizer/tests/ground.rs

```diff
@@ -0,0 +1,286 @@
+        multival: false,
+        ..Default::default()
+    });
+    add_attribute(&mut schema, 66, Attribute {
+        value_type: ValueType::Ref,
+        multival: true,
+        ..Default::default()
+    });
+    add_attribute(&mut schema, 67, Attribute {
+        value_type: ValueType::String,
+        multival: true,
+        ..Default::default()
+    });
+    add_attribute(&mut schema, 68, Attribute {
+        value_type: ValueType::Long,
+        multival: false,
+        ..Default::default()
+    });
+    add_attribute(&mut schema, 69, Attribute {
+        value_type: ValueType::Long,
+        multival: false,
+        ..Default::default()
+    });
+    schema
+}
+
+fn bails(schema: &Schema, input: &str) -> Error {
+    let parsed = parse_find_string(input).expect("parse failed");
+    algebrize(schema.into(), parsed).expect_err("algebrize should have failed")
+}
+
+fn alg(schema: &Schema, input: &str) -> ConjoiningClauses {
+    let parsed = parse_find_string(input).expect("parse failed");
+    algebrize(schema.into(), parsed).expect("algebrize failed").cc
+}
+
+#[test]
+fn test_ground_doesnt_bail_for_type_conflicts() {
+    // We know `?x` to be a ref, but we're attempting to ground it to a Double.
+    // The query can return no results.
+    let q = r#"[:find ?x :where [?x :foo/knows ?p] [(ground 9.95) ?x]]"#;
+    let schema = prepopulated_schema();
+    let cc = alg(&schema, &q);
+    assert!(cc.empty_because.is_some());
+}
+
+#[test]
+fn test_ground_tuple_fails_impossible() {
+    let q = r#"[:find ?x :where [?x :foo/knows ?p] [(ground [5 9.95]) [?x ?p]]]"#;
+    let schema = prepopulated_schema();
+    let cc = alg(&schema, &q);
+    assert!(cc.empty_because.is_some());
+}
+
+#[test]
+fn test_ground_scalar_fails_impossible() {
+    let q = r#"[:find ?x :where [?x :foo/knows ?p] [(ground true) ?p]]"#; 
+    let schema = prepopulated_schema();
+    let cc = alg(&schema, &q);
+    assert!(cc.empty_because.is_some());
+}
+
+#[test]
+fn test_ground_coll_skips_impossible() {
+    // We know `?x` to be a ref, but we're attempting to ground it to a Double.
+    // The query can return no results.
+    let q = r#"[:find ?x :where [?x :foo/knows ?p] [(ground [5 9.95 11]) [?x ...]]]"#;
+    let schema = prepopulated_schema();
+    let cc = alg(&schema, &q);
+    assert!(cc.empty_because.is_none());
+    assert_eq!(cc.computed_tables[0], ComputedTable::NamedValues {
+        names: vec![Variable::from_valid_name("?x")],
+        values: vec![TypedValue::Ref(5), TypedValue::Ref(11)],
+    });
+}
+
+#[test]
+fn test_ground_coll_fails_if_all_impossible() {
+    let q = r#"[:find ?x :where [?x :foo/knows ?p] [(ground [5.1 5.2]) [?p ...]]]"#;
+    let schema = prepopulated_schema();
+    let cc = alg(&schema, &q);
+    assert!(cc.empty_because.is_some());
+}
+
+#[test]
+fn test_ground_rel_skips_impossible() {
+    let q = r#"[:find ?x :where [?x :foo/knows ?p] [(ground [[8 "foo"] [5 7] [9.95 9] [11 12]]) [[?x ?p]]]]"#;
```

Comment:
I still do require homogeneous columns.

These two features are overlapping: we quietly eliminate rows that can't generate bindings, for whatever reason (out-of-range value, type mismatch against the rest of the query, ident doesn't resolve). If we're left with a column that has more than one possible type — e.g., if there's no type imposed by the rest of the query — then we bail.

We bail because we haven't implemented this feature: we have no way of representing a set of bindings of different types for a single variable. We quietly fail when a query is valid but can't return results in _this_ store with _this_ schema.